ADR-0000 Introduce ADR process (#21947)
opf/openproject: docs/decisions/README.md

# Architecture Decision Records (ADRs)

This project uses **Architecture Decision Records (ADRs)** to document important technical and architectural decisions. ADRs ensure that we preserve context, reasoning, and trade‑offs so that current and future developers understand _why_ the system looks the way it does — not only _how_ it works.

An ADR becomes part of the project’s architecture documentation and is authoritative once accepted and merged.

---

## Where ADRs live

All ADRs are stored in the repository under:

```

> /docs/decisions/

Reviewer: Side question: Will this lead to ADRs becoming part of our website automatically? Is this intended/unintended?

Author: I guess @as-op needs to answer this? I don't mind it becoming part of the developer docs. But it is not needed imho.

```

The ADR template is located at:

```
/docs/decisions/adr-template.md
```

---

## When an ADR is required

You must create an ADR when making a decision that is:

- Architectural or system‑wide
- Difficult to reverse
- Expensive to change later
- Introducing or replacing a technology
- Introducing a new RubyGem or NPM package when it is not trivial and introduce new usage patterns (i.e. Angular, React, dry-rb, etc.)
- Introducing a new database or service that needs to be included in the deployment process
- Introducing dependencies on external services
- Affecting APIs or system boundaries

> - Influencing how multiple teams work

Reviewer: I can't find this bullet point in ADR-0000 under "ADRs will be required for significant architectural decisions, such as". How does this ADR creation criteria get here?

Author: added

Typical examples:

- Introducing a framework or library
- Database technology decisions
- Authentication or authorization approach
- Service boundaries or modularization
- Communication patterns (events vs REST, sync vs async)

> - Deployment or hosting strategy

Reviewer 2: Not too relevant as this is just an example, but I doubt we'll document deployment strategies in an ADR

Reviewer 3: ```suggestion
- Frontend rendering approach (server-rendered vs. single-page app)
```

Reviewer 3: (a more realistic example, hopefully!)

If unsure: **create an ADR**. It is better to document one decision too many than one too few.

---

## Creating a New ADR (Step‑by‑Step)

1. **Copy the template**

   Copy the file:

   ```
   /docs/decisions/adr-template.md
   ```

2. **Choose the next number**

>    ADRs are numbered sequentially.

Reviewer: This question might sound stupid, but how do we handle race conditions?

~~I.e. since ADRs are accepted in pull requests, it's easily possible that the latest ADR in `dev` is 0000, but there are already two PRs that each want to introduce a new ADR.~~ edit: Just now read that the answer is "coordinate", adjacent question below remains valid as far as I can tell.

Adjacent question: Do rejected ADRs get a number and a place in the docs as well? (I can see good and bad for both ways)

Author: Note that there is no `rejected` state. When an ADR is suggested and the team decides against it, the PR is closed and it is not added to the repo.

Author: As for the conflict ... We will not be adding 100s of ADRs. If two ADRs come in at the same time, I count on mature developers agreeing that one has to rename theirs.

   Format:

   ```
   adr-XXXX-short-title.md
   ```

   Example:

   ```
   adr-0005-use-dry-monads-for-services.md
   ```

   Use the next available number in the directory. If a different decision is being made in parallel, coordinate to avoid conflicts.

3. **Fill out the ADR**

   Fill in as many sections as possible. It is acceptable if some sections evolve during discussion — the pull request is part of the decision process.

4. **Open a Pull Request**

   Create a pull request containing only the ADR (and related material if necessary).

5. **Notify the team**

   Post a message in the **Development Matrix channel** announcing the proposed ADR.

6. **Request collaboration**

   If specific developers have expertise, explicitly request their input and collaboration.

> 7. **Add to Dev Weekly agenda**

Reviewer: Something that I notice: Some of the process parts are restricted to OpenProject GmbH employees. I think practically this is fine, since I've not seen any outside contribution/contributor yet that had a significant enough impact that they'd ever need an ADR for their contributions.

Still it feels like a violation of the **open** aspect in open source, that certain parts of the development process for this tool are by design reserved to employees.

Author: I think it is us setting the agenda here.

Reviewer: Not sure what you mean, maybe I wasn't clear enough about my concern:

Right now the process requires mentioning a private matrix channel and putting the topic up on an agenda of a private meeting. This excludes the public from this kind of change on a process level.

As I said, this is probably not a _practical_ problem, since I've not seen any outside contributions of a magnitude that they would require (or be interested in) changing ADRs. Still we should be aware that this is a barrier we are adding here.

Author: I think we should design the process so that it works for us, the company. If we ever see outside contribution in a relevant mass, where it would make sense that people would attend our dev discussion meetings, we will have to set up a lot of different processes anyway. I think tackling this now might be premature optimization for a case that will likely never happen.

If it comes to this, that an outside contributor is heavily invested in the process, we can invite them to meetings.

Reviewer 3: Noted — keeping internal steps for now; the public PR remains the open record. Can revisit if external ADR contributions actually materialise. ❓

   Add an agenda item to the next [weekly dev meeting](https://community.openproject.org/projects/development/recurring_meetings/2).

8. **Discuss and refine**

   Use pull request comments and the meeting discussion to refine the ADR.

9. **Confirmation steps**

   If the ADR includes Confirmation steps, ensure that you create a work package within OpenProject to implement these steps as soon as possible. Link the ticket in the Pull Request description.

10. **Approval**

>     The ADR must be approved by **at least one developer from each team**.

Reviewer: I think the consensus process that ADRs follow is an important aspect of how ADRs work. However, I can't find it described in ADR-0000.

Shouldn't ADR-0000 be the authoritive source for how ADRs work?

Author: That's very reasonable, will add that.

The ADR0000 should be the authority on what should be part of an ADR. The readme specifies a bit of the surrounding processes (like presenting it in the dev weekly). But I think those can and should be easily changeable.

>     The ADR must be approved by **at least one developer from each team**.

Reviewer 2: I don't think ADRs need approval from at least one member of every team. That might be a good indicator for approval, but not a requirement. I'd rather phrase it like every team needs to have been given the opportunity to review the ADR, and move approval to someone like a consensus between architects.

Reviewer 3: ```suggestion
    Be accepted through a pull request that gives every team the opportunity to review, with final approval by consensus among the team leads, then merged into the `dev` branch.
```

11. **Merge**

    Once approvals are present, the ADR can be merged.

---

## After Merge

When an ADR is merged:

- The decision becomes **mandatory** for the entire development team
- All developers must follow it
- Pull request reviewers should actively point out violations

> - If the ADR contained Confirmation steps, these should be implemented as soon as possible

Reviewer 3: ```suggestion
- The ADR records a shared team decision and establishes the preferred default approach.
- Developers should apply the spirit and intent of the decision, using their professional judgment where the ADR does not (or cannot) give explicit guidance.
- Code reviewers should surface deviations and uncertainties to ensure the decision is being understood and applied consistently.
- If the ADR defines Confirmation steps, these should be implemented as soon as is reasonably practical.
```

There should be an "escape hatch" for when ADR decisions might (occasionally) need to be overridden. I'm not sure if we need a separate section for this, but it could look like:

> ## Exceptions
> 
> - Exceptions are allowed when the ADR is impractical, incomplete, or counterproductive in a specific context.
> - Any exception should be explicit and visible (for example: noted in the pull request, code comments, or follow-up ADR).
> - Exceptions and recurring friction are treated as input to the feedback loop, and may result in clarifying, amending, or superseding the ADR.
> - An ADR should never be used as a disciplinary tool; its purpose is alignment, not enforcement.
>

Author: I kind of disagree. When we give escape routes, we are talking about guidelines and not rules. For example when we decide to use Angular as our Frontend Framework, a developer should not be allowed to just add a React component and comment in the PR with "made more sense here". In this case, the ADR that forces us to use Angular should be archived and superseded by another one that allows Angular and React for certain cases.

Reviewer: > its purpose is alignment, not enforcement

While I like this, it feels like a slippery slope to me. Essentially we can ignore ADRs whenever we don't like their impact.

It's a bit like "asking for forgiveness" instead of "asking for permission".

(I am not saying we shouldn't allow exceptions, but I am not sure whether we should formally put the option for exceptions in)

Reviewer 3: > (I am not saying we shouldn't allow exceptions, but I am not sure whether we should formally put the option for exceptions in)

If people do find it difficult to adhere for whatever reason, then it's much better if those reasons are visible and documented. We don't want devs slipping in changes surreptitiously  because they find one of our ADR decisions too rigid or the process too arduous. We do want our process to reflect reality!

> It's a bit like "asking for forgiveness" instead of "asking for permission".

FWIW, when in doubt, I always opt for the first.

Reviewer 3: > a developer should not be allowed to just add a React component and comment in the PR with "made more sense here". 

That's clearly not an acceptable level of documentation! I've actually opened a couple PRs that use Primer React components - primarily for discussion / Proof of Concept. I wouldn't want that ability to experiment to be curtailed!

The onus would be on me to elaborate a bit. Perhaps:

"This is an experiment that investigates exposing React components as custom elements in order to take advantage of the improvements only currently available in the Primer implementation.

@akabiru

Author: I think it will be a question of how detailed we will go with ADRs. I don't think we will have tons of them covering every little detail of how we write code. Just important ones, like introducing new technologies (like Primer View Components for example).

Actually, Primer View Components are a very good example. Since their introduction, we have not added new pages to our code that use SPOT. Whenever we touched something we thought about migrating it to Primer when it was feasible. So this is the perfect example of what would be handled in an ADR and how it is enforced.

**If** we did an ADR that enforces for example Angular as our frontend technology, then I personally would say, you can experiment with React, sure, but if you want to merge it into the codebase, the ADR has to be adjusted before we do this. 

Same thing goes for if we say in an ADR that Service objects are done using our Service class, then it is **forbidden** to introduce something like dry-monads into the codebase until the ADR has been replaced and explicitly allows for either both or only the new one.

One can experiment in PRs and make a case for why the ADR needs to be changed.

But we all know too well, if we start violating something like this we agreed on in one place with the intent "let's just add this one React thing here, see how it goes, and in a few weeks we will remove it" ... That then gets fogotten and stays in there forever until somebody comes along and is wondering why there is React in here, if we clearly state in an ADR that we do not use React.

Reviewer: I am with Klaus on this one to be honest. Otherwise ADRs risk to become pointless. However, let's also be clear that it's in the end on the ADR to allow for certain exclusions.

I find it hard to discuss @myabc's use of React components right now, because an ADR forbidding to do so, doesn't exist. What would our ADR look like? Does it say "we **only** use tool X" or does it say "we primiarily use tool X" or "in this specified area we use tool X"?

If we don't treat ADRs as binding agreements among developers, then I am not sure why we should do them.

> An ADR should never be used as a disciplinary tool; its purpose is alignment, not enforcement.

And maybe to elaborate on terms such as "binding agreement", "disciplinary tool" and "enforcement". So far I didn't read anything about fines to be paid or jail time to be incurred, if you violate an ADR. So I don't see any enforcing functions included.
It's a gentlemen's agreement that we do not violate the rules we set out to play by and I think there will be plenty of violating those rules happening even without us already writing down every possible way that we do not want to follow them beforehand.

Author: > So I don't see any enforcing functions included.

Well, it states in the template that we should however possible use "enforcement" tools like Rubocop, Danger, etc. While those do not fine developers, there should be as much enforcement as possible.

For example, when we add a rubocop rule that will enforce an ADR, there should at least be a discussion in the pull request, when that rule is explicitly disabled for a file that was added.

Reviewer 3: > Does it say "we only use tool X" or does it say "we primiarily use tool X" or "in this specified area we use tool X"?

Primer View Component is the default. Use Primer's View Compnent implementation unless you have a very good reason not to.

I guess this could be codified in the relevant ADR rather than ADR 0001.

This section should allow a reader to understand the decision without prior knowledge of the discussion.

---

### Decision Drivers

List the forces influencing the decision.

Examples:

- performance requirements
- scalability
- team expertise
- operational cost
- compliance
- maintainability

These are the evaluation criteria for the options.

---

### Considered Options

List realistic alternatives that were actually evaluated.

Include at least two options whenever possible.

Avoid strawman options.

---

### Decision Outcome

State the chosen option and _why it won_.

This is the most important section.

Someone should be able to read only this section and understand the conclusion.

---

### Consequences

Document what the decision causes.

Include both:

- positive effects
- negative effects

Every real decision has trade‑offs. Documenting the downsides prevents future confusion.

---

### Confirmation

Describe how we ensure the decision is followed.

Examples:

- code review checks
- automated tests
- linters
- architecture validation tools
- RuboCop rules
- Dangerfile checks
- Custom actions

---

### Pros and Cons of the Options

> Explain the reasoning process for each option.
> 
> This section answers the future developer’s question:

Author: ```suggestion
Explain the reasoning process for each option.

When the ADR is about introducing a package or a new coding pattern, show some code examples for each option so we can compare how the newly introduced pattern will be used.

This section answers the future developer’s question:
```

> > Why didn’t we choose the other obvious solution?

Author: Here it might be an option that we also provide code samples how the other implementations would look like. I did not add it to the README yet, but would probably be interesting.

Reviewer 4: This could be useful exactly in the cases of introducing new library / pattern.

Reviewer 3: [This suggestion should cover this already](https://github.com/opf/openproject/pull/21947/changes#r2789033352)

---

### More Information

Optional supporting material:

- links
- meeting outcomes
- follow‑up work
- revisit conditions

---

## Review Responsibilities

All developers share responsibility for maintaining architectural consistency.

During PR reviews you should:

- check whether an ADR applies
- request an ADR if a major decision is introduced
- point out ADR violations

---

## Goal of ADRs

> ADRs are not bureaucracy.

Reviewer: Said every bureaucrat about every bureaucratic process ever. Most processes were born out of good intentions, some people even pave roads with good intentions.

I still think that we should try it out and see whether they help, but reading a line like this is literal rage bait for me xD

Reviewer 3: ```suggestion
ADRs exist to save time, not to add process.
```

- Deployment or hosting strategy

If unsure: **create an ADR**. It is better to document one decision too many than one too few.

---

## Creating a New ADR (Step‑by‑Step)

1. **Copy the template**

   Copy the file:

   ```
   /docs/decisions/adr-template.md
   ```

2. **Choose the next number**

   ADRs are numbered sequentially.

   Format:

   ```
   adr-XXXX-short-title.md
   ```

   Example:

   ```
   adr-0005-use-dry-monads-for-services.md
   ```

   Use the next available number in the directory. If a different decision is being made in parallel, coordinate to avoid conflicts.

3. **Fill out the ADR**

   Fill in as many sections as possible. It is acceptable if some sections evolve during discussion — the pull request is part of the decision process.

4. **Open a Pull Request**

   Create a pull request containing only the ADR (and related material if necessary).

5. **Notify the team**

   Post a message in the **Development Matrix channel** announcing the proposed ADR.

6. **Request collaboration**

   If specific developers have expertise, explicitly request their input and collaboration.

7. **Add to Dev Weekly agenda**

>    Add an agenda item to the next [weekly dev meeting](https://community.openproject.org/projects/development/recurring_meetings/2).

Reviewer 3: ```suggestion
   Add an agenda item to the next [weekly dev meeting](https://community.openproject.org/projects/development/recurring_meetings/2). Steps relying on internal channels (this one and "Notify the team") are carried out by a maintainer on behalf of external contributors; the pull request stays the public, authoritative record of the discussion.
```
